File: 0068-text-justification/0068-text-justification.py
class Solution:
  def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
    def leftJustify(words: List[str], maxWidth: int) -> str:
      ans = " ".join(words)
      tail = " "*(maxWidth-len(ans))
      ans += tail
      return ans
    
    def justifyLine(words: List[str], maxWidth: int) -> str:
      if len(words) == 1:
        return leftJustify(words, maxWidth)
      wordsLen = sum(len(word) for word in words)
      spaces = maxWidth-wordsLen
      ans = ""
      for i in range(len(words)-1):
        ans += words[i]
        base = spaces//(len(words)-1)
        leftover = 1 if i < spaces%(len(words)-1) else 0
        ans += " "*(base+leftover)
      ans += words[-1]
      return ans
    
    lines = []
    l = r = 0
    minWidth = len(words[0])
    while (l < len(words)):
      if minWidth == maxWidth:
        lines.append(" ".join(words[l:r+1]))
        l = r+1
        r = l
        minWidth = len(words[l]) if l < len(words) else 0
      elif minWidth > maxWidth:
        minWidth -= len(words[r])
        minWidth -= 1
        r -= 1
        lines.append(justifyLine(words[l:r+1], maxWidth))
        l = r+1
        r = l
        minWidth = len(words[l]) if l < len(words) else 0
      else:
        r += 1
        if r < len(words):
          minWidth += len(words[r])
          minWidth += 1
        else:
          lines.append(leftJustify(words[l:], maxWidth))
          break
    return lines

File: 0068-text-justification/0068-text-justification.ts
function fullJustify(words: string[], maxWidth: number): string[] {
  function leftJustify(words: string[], maxWidth: number): string {
    let ans = words.join(" ")
    const tail: string = " ".repeat(maxWidth - ans.length)
    return ans + tail
  }
  
  function justifyLine(words: string[], maxWidth: number): string {
    if (words.length === 1) return leftJustify(words, maxWidth)
    const wordsLen: number = words.reduce((sum, word) => sum + word.length, 0)
    let spaces: number = maxWidth - wordsLen
    const wordsCount: number = words.length
    let ans: string = ""
    let base: number
    let leftover: number
    for (let i = 0; i < wordsCount - 1; i++) {
      ans += words[i]
      base = Math.floor(spaces / (wordsCount - 1))
      leftover = i < spaces % (wordsCount - 1) ? 1 : 0
      ans += " ".repeat(base + leftover)
    }
    ans += words.at(-1)
    return ans
  }
  
  const lines = []
  let [l, r, minWidth] = [0, 0, words[0].length]
  while (l < words.length) {
    if (minWidth === maxWidth) {
      lines.push(words.slice(l, r + 1).join(" "))
      l = r + 1
      r = l
      minWidth = l < words.length ? words[l].length : 0
    } else if (minWidth > maxWidth) {
      minWidth -= words[r].length
      minWidth--
      r--
      lines.push(justifyLine(words.slice(l, r + 1), maxWidth))
      l = r + 1
      r = l
      minWidth = l < words.length ? words[l].length : 0
    } else if (minWidth < maxWidth) {
      r++
      if (r < words.length) {
        minWidth += words[r].length
        minWidth++
      } else {
        lines.push(leftJustify(words.slice(l), maxWidth))
        break
      }
    }
  }
  return lines
}